crates/trippy-tui/src/locale.rs

```diff
@@ -1,43 +1,125 @@
+use arc_swap::ArcSwap;
+use itertools::Itertools;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::sync::{Arc, OnceLock};
+
 const FALLBACK_LOCALE: &str = "en";
 
 /// Set the locale for the application.
 ///
-/// If the given locale is `None` or unsupported, the system locale is tried. If the system locale
-/// is not supported, the fallback locale is used.
+/// If the given locale is `None` the system locale is tried. If the system locale cannot be
+/// determined then the fallback locale is used.
 ///
-/// In both cases, the language part of the locale is used if the full locale is not supported.
-pub fn set_locale(locale: Option<&str>) {
-    if let Some(locale) = locale {
-        set_locale_inner(locale);
-    } else if let Some(locale) = sys_locale::get_locale().as_ref() {
-        set_locale_inner(locale);
+/// In all cases, the language part of the locale is used if the full locale is not supported.
+pub fn set_locale(locale: Option<&str>) -> String {
+    let new_locale = if let Some(locale) = locale {
+        String::from(locale)
+    } else if let Some(locale) = sys_locale::get_locale() {
+        locale
     } else {
-        set_locale_inner(FALLBACK_LOCALE);
-    }
+        String::from(FALLBACK_LOCALE)
+    };
+    set_locale_inner(new_locale)
 }
 
-/// Get the current locale.
-pub fn locale() -> String {
-    rust_i18n::locale().to_string()
+/// Get the available locales.
+pub fn available_locales() -> Vec<&'static str> {
```

Comment:
Will give the sorted list of unique locales available, without assuming any particular item exists.